core/src/main/java/org/apache/gravitino/cache/CaffeineEntityCache.java

```diff
@@ -0,0 +1,493 @@
+   * entry expires, it will call this method.
+   *
+   * @param key The key of the expired entity
+   */
+  @Override
+  protected void invalidateExpiredItem(EntityCacheKey key) {
+    withLock(
+        () -> {
+          cacheIndex.remove(key.toString());
+        });
+  }
+
+  /**
+   * Syncs the entities to the cache, if cache is too big and can not put it to the cache, then it
+   * will remove and cacheIndex will not be updated.
+   *
+   * @param key The key of the entities.
+   * @param newEntities The new entities to sync to the cache.
+   */
+  private void syncEntitiesToCache(EntityCacheKey key, List<Entity> newEntities) {
+    List<Entity> existingEntities = cacheData.getIfPresent(key);
+
+    if (existingEntities != null && key.relationType() != null) {
+      List<Entity> merged = new ArrayList<>(existingEntities);
+      merged.addAll(newEntities);
+
+      cacheData.put(key, merged);
+    } else {
+      cacheData.put(key, newEntities);
+    }
+
+    if (cacheData.policy().getIfPresentQuietly(key) != null) {
+      cacheIndex.put(key.toString(), key);
+    }
+  }
+
+  /**
+   * Returns a new instance of Caffeine cache builder.
+   *
+   * @param cacheConfig The cache configuration
+   * @param <KEY> The key type
+   * @param <VALUE> The value type
+   * @return The new instance of Caffeine cache builder
+   */
+  @SuppressWarnings("unchecked")
+  private <KEY, VALUE> Caffeine<KEY, VALUE> newBaseBuilder(Config cacheConfig) {
+    Caffeine<Object, Object> builder = Caffeine.newBuilder();
+
+    if (cacheConfig.get(Configs.CACHE_WEIGHER_ENABLED)) {
+      builder.maximumWeight(EntityCacheWeigher.getMaxWeight());
+      builder.weigher(EntityCacheWeigher.getInstance());
+    } else {
+      builder.maximumSize(cacheConfig.get(Configs.CACHE_MAX_ENTRIES));
+    }
+
+    if (cacheConfig.get(Configs.CACHE_EXPIRATION_ENABLED)) {
+      builder.expireAfterWrite(
+          cacheConfig.get(Configs.CACHE_EXPIRATION_TIME), TimeUnit.MILLISECONDS);
+    }
+
+    if (cacheConfig.get(Configs.CACHE_STATUS_ENABLED)) {
+      builder.recordStats();
+    }
+
+    return (Caffeine<KEY, VALUE>) builder;
+  }
+
+  /**
+   * Invalidates the entities by the given cache key.
+   *
+   * @param identifier The identifier of the entities to invalidate
+   */
+  private void invalidateEntities(NameIdentifier identifier) {
+    Iterable<EntityCacheKey> entityKeysToRemove =
+        cacheIndex.getValuesForKeysStartingWith(identifier.toString());
+
+    cacheData.invalidateAll(entityKeysToRemove);
+    entityKeysToRemove.forEach(key -> cacheIndex.remove(key.toString()));
+  }
+
+  /**
+   * Runs the given action with the lock.
+   *
+   * @param action The action to run with the lock
+   */
+  private void withLock(Runnable action) {
+    try {
+      opLock.lockInterruptibly();
+      try {
+        action.run();
+      } finally {
+        opLock.unlock();
+      }
+    } catch (InterruptedException e) {
+      Thread.currentThread().interrupt();
```

Comment:
Shall we interrupt the current thread, if so then this entity cache will not work any more, right?